ksql-engine/src/main/java/io/confluent/ksql/planner/plan/FlatMapNode.java

```diff
@@ -130,14 +132,23 @@ private LogicalSchema buildLogicalSchema(final LogicalSchema inputSchema) {
 
     final ConnectToSqlTypeConverter converter = SchemaConverters.connectToSqlConverter();
 
+    final ExpressionTypeManager expressionTypeManager = new ExpressionTypeManager(
+        inputSchema, functionRegistry);
+
     // And add new columns representing the exploded values at the end
     for (int i = 0; i < analysis.getTableFunctions().size(); i++) {
+      final FunctionCall functionCall = analysis.getTableFunctions().get(i);
       final KsqlTableFunction tableFunction =
           UdtfUtil.resolveTableFunction(functionRegistry,
-              analysis.getTableFunctions().get(i), inputSchema
+              functionCall, inputSchema
           );
       final ColumnName colName = ColumnName.synthesisedSchemaColumn(i);
-      final SqlType fieldType = converter.toSqlType(tableFunction.getReturnType());
+      final List<Schema> argTypes = new ArrayList<>(functionCall.getArguments().size());
+      for (Expression expression : functionCall.getArguments()) {
```

Comment:
can't we just call `expressionTypeManager.getExpressionSqlType(tableFunction)`? It already has the logic to call `getReturnType` on the resolved args.

Reply:
It's odd, but I did it this way because `public Schema getExpressionSchema(final Expression expression) {` is deprecated and I wanted to avoid deprecation warnings.

Reply:
is there a good reason why it's deprecated? Is it because it returns the `Schema`? I think we should be able to clean that up

Reply:
I'm assuming it's been marked that way as use of Schema is not frowned up on favour of SqlType. Not sure there is much I can do here.

Reply:
Have simplified after discussion with @agavra